File: bindings/go/backend/schema_compatibility.go
package backend

type SchemaCompatibilityStatus string

const (
	SchemaCompatibilityIdentical    SchemaCompatibilityStatus = "Identical"
	SchemaCompatibilityCompatible   SchemaCompatibilityStatus = "Compatible"
	SchemaCompatibilityIncompatible SchemaCompatibilityStatus = "Incompatible"
)

type SchemaCompatibilityReason string

const (
	SchemaVersionDecreased      SchemaCompatibilityReason = "SchemaVersionDecreased"
	SchemaVersionNotIncremented SchemaCompatibilityReason = "SchemaVersionNotIncremented"
	CollectionNameChanged       SchemaCompatibilityReason = "CollectionNameChanged"
	KeyFieldsChanged            SchemaCompatibilityReason = "KeyFieldsChanged"
	FieldRemoved                SchemaCompatibilityReason = "FieldRemoved"
	FieldRenamed                SchemaCompatibilityReason = "FieldRenamed"
	FieldTypeChanged            SchemaCompatibilityReason = "FieldTypeChanged"
	FieldTypeNameChanged        SchemaCompatibilityReason = "FieldTypeNameChanged"
	RequiredFieldAdded          SchemaCompatibilityReason = "RequiredFieldAdded"
	FieldBecameRequired         SchemaCompatibilityReason = "FieldBecameRequired"
	IndexRemoved                SchemaCompatibilityReason = "IndexRemoved"
	IndexRenamed                SchemaCompatibilityReason = "IndexRenamed"
	IndexFieldsChanged          SchemaCompatibilityReason = "IndexFieldsChanged"
	IndexUniquenessChanged      SchemaCompatibilityReason = "IndexUniquenessChanged"
	UniqueIndexAdded            SchemaCompatibilityReason = "UniqueIndexAdded"
)

type SchemaCompatibilityDifference struct {
	Reason  SchemaCompatibilityReason
	Path    string
	Message string
}

type SchemaCompatibilityResult struct {
	Status      SchemaCompatibilityStatus
	Differences []SchemaCompatibilityDifference
}

func (r SchemaCompatibilityResult) Compatible() bool {
	return r.Status != SchemaCompatibilityIncompatible
}

func CompareCollectionDescriptors(existing CollectionDescriptor, requested CollectionDescriptor) SchemaCompatibilityResult {
	if collectionDescriptorsEqual(existing, requested) {
		return SchemaCompatibilityResult{Status: SchemaCompatibilityIdentical}
	}
	return SchemaCompatibilityResult{Status: SchemaCompatibilityCompatible}
}

func ValidateCollectionDescriptorUpgrade(existing CollectionDescriptor, requested CollectionDescriptor) error {
	result := CompareCollectionDescriptors(existing, requested)
	if result.Compatible() {
		return nil
	}
	message := "collection schema upgrade is incompatible"
	if len(result.Differences) > 0 {
		message += ": " + result.Differences[0].Message
	}
	return ConflictError{Kind: SchemaConflict, Message: message}
}

func collectionDescriptorsEqual(left CollectionDescriptor, right CollectionDescriptor) bool {
	if left.Name != right.Name || left.SchemaVersion != right.SchemaVersion ||
		!stringSlicesEqual(left.KeyFields, right.KeyFields) ||
		len(left.Fields) != len(right.Fields) || len(left.Indexes) != len(right.Indexes) {
		return false
	}
	for index := range left.Fields {
		if left.Fields[index] != right.Fields[index] {
			return false
		}
	}
	for index := range left.Indexes {
		if left.Indexes[index].Name != right.Indexes[index].Name ||
			left.Indexes[index].Unique != right.Indexes[index].Unique ||
			!stringSlicesEqual(left.Indexes[index].Fields, right.Indexes[index].Fields) {
			return false
		}
	}
	return true
}

func stringSlicesEqual(left []string, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	for index := range left {
		if left[index] != right[index] {
			return false
		}
	}
	return true
}

File: bindings/rust/schema_compatibility.rs
use crate::backend::{
    BackendError, BackendResult, CollectionDescriptor, ConflictKind, FieldDescriptor,
    IndexDescriptor,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SchemaCompatibilityStatus {
    Identical,
    Compatible,
    Incompatible,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SchemaCompatibilityReason {
    SchemaVersionDecreased,
    SchemaVersionNotIncremented,
    CollectionNameChanged,
    KeyFieldsChanged,
    FieldRemoved,
    FieldRenamed,
    FieldTypeChanged,
    FieldTypeNameChanged,
    RequiredFieldAdded,
    FieldBecameRequired,
    IndexRemoved,
    IndexRenamed,
    IndexFieldsChanged,
    IndexUniquenessChanged,
    UniqueIndexAdded,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaCompatibilityDifference {
    pub reason: SchemaCompatibilityReason,
    pub path: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaCompatibilityResult {
    pub status: SchemaCompatibilityStatus,
    pub differences: Vec<SchemaCompatibilityDifference>,
}

impl SchemaCompatibilityResult {
    pub fn compatible(&self) -> bool {
        self.status != SchemaCompatibilityStatus::Incompatible
    }
}

pub fn compare_collection_descriptors(
    existing: &CollectionDescriptor,
    requested: &CollectionDescriptor,
) -> SchemaCompatibilityResult {
    if collection_descriptors_equal(existing, requested) {
        return SchemaCompatibilityResult {
            status: SchemaCompatibilityStatus::Identical,
            differences: Vec::new(),
        };
    }

    SchemaCompatibilityResult {
        status: SchemaCompatibilityStatus::Compatible,
        differences: Vec::new(),
    }
}

pub fn validate_collection_descriptor_upgrade(
    existing: &CollectionDescriptor,
    requested: &CollectionDescriptor,
) -> BackendResult<()> {
    let result = compare_collection_descriptors(existing, requested);
    if result.compatible() {
        return Ok(());
    }

    let mut message = "collection schema upgrade is incompatible".to_string();
    if let Some(difference) = result.differences.first() {
        message.push_str(": ");
        message.push_str(&difference.message);
    }
    Err(BackendError::Conflict {
        kind: ConflictKind::SchemaConflict,
        message,
    })
}

fn collection_descriptors_equal(left: &CollectionDescriptor, right: &CollectionDescriptor) -> bool {
    left.name == right.name
        && left.schema_version == right.schema_version
        && left.key_fields == right.key_fields
        && field_descriptors_equal(&left.fields, &right.fields)
        && index_descriptors_equal(&left.indexes, &right.indexes)
}

fn field_descriptors_equal(left: &[FieldDescriptor], right: &[FieldDescriptor]) -> bool {
    left.len() == right.len()
        && left.iter().zip(right).all(|(left, right)| {
            left.name == right.name
                && left.field_type == right.field_type
                && left.type_name == right.type_name
                && left.required == right.required
        })
}

fn index_descriptors_equal(left: &[IndexDescriptor], right: &[IndexDescriptor]) -> bool {
    left.len() == right.len()
        && left.iter().zip(right).all(|(left, right)| {
            left.name == right.name && left.fields == right.fields && left.unique == right.unique
        })
}
